Add autoupdate agent type validations

hugoShaka opened the pull request:
This PR adds the validation logic for agent autoupdate fields that were introduced in https://github.com/gravitational/teleport/pull/47666

Depends on https://github.com/gravitational/teleport/pull/47830

Part of [RFD 0184](https://github.com/gravitational/teleport/pull/47126).

hugoShaka committed:
Add autoupdate agent validations

hugoShaka committed:
Add AutoUpdateAgentRollout constants

sclevine commented on api/types/autoupdate/constants.go:
```diff
@@ -0,0 +1,29 @@
+package autoupdate
+
+const (
+	// ToolsUpdateModeEnabled enables client tools automatic updates.
+	ToolsUpdateModeEnabled = "enabled"
+	// AgentsUpdateModeEnabled enabled agent automatic updates.
+	AgentsUpdateModeEnabled
+	// ToolsUpdateModeDisabled disables client tools automatic updates.
+	ToolsUpdateModeDisabled = "disabled"
+	// AgentsUpdateModeDisabled disables agent automatic updates.
+	AgentsUpdateModeDisabled
+	// AgentsUpdateModeSuspended temporarily suspends agent automatic updates.
+	AgentsUpdateModeSuspended = "suspended"
```
Not strongly opinionated, but I'd favor repeating the string constants and defining separate groupings for Tools vs. Agents. Too clever for me 🙂

hugoShaka commented on api/types/autoupdate/constants.go:
```diff
@@ -0,0 +1,29 @@
+package autoupdate
```
```suggestion
/*
 * Teleport
 * Copyright (C) 2024  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package autoupdate
```

hugoShaka commented on api/types/autoupdate/rollout_test.go:
```diff
@@ -0,0 +1,130 @@
+package autoupdate
```
```suggestion
/*
 * Teleport
 * Copyright (C) 2024  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package autoupdate
```

hugoShaka commented on api/types/autoupdate/utils.go:
```diff
@@ -0,0 +1,52 @@
+package autoupdate
```
```suggestion
/*
 * Teleport
 * Copyright (C) 2024  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package autoupdate
```

vapopov commented on api/types/autoupdate/rollout_test.go:
```diff
@@ -0,0 +1,130 @@
+)
+
+// TestNewAutoUpdateConfig verifies validation for AutoUpdateConfig resource.
+
```
```suggestion

```

hugoShaka committed:
Fix autoupdate API licenses

Teleport's `api/` and `integrations/` should be Apache-licensed.

Only the main teleport process should be licenses under AGPLv3.

hugoShaka committed:
address feedback

hugoShaka commented:
I also fixed all the `api/types/autoupdate` license headers. `api` is Apache-licensed, unlike the rest of the Telpeort codebase which is AGPLv3.
